Initial Unit Testing structure for Client API

Mierdin opened the pull request:
A few changes:

- (main change) - Added unit testing structure for the "client" portion of the ToDD API. Provided example in agents_test.go
- Broke out config structs so that they can be better leveraged during testing
- refactored agents.go to return an error condition, and a slice of Agent definitions.

vCabbage commented on api/client/agents_test.go:
```diff
@@ -0,0 +1,95 @@
+	}))
+	defer ts.Close()
+
+	agentsUrl := fmt.Sprintf("%s/v1/agent", ts.URL)
```
Doesn't look like the "/v1/agent" part is being used, just gets removed below.

Mierdin replied:
Ah, this was a holdover from an earlier approach to mocking the API. Removing

vCabbage commented on api/client/agents_test.go:
```diff
@@ -0,0 +1,95 @@
+	err, agents := capi.Agents(
+		map[string]string{
+			"host": strings.Split(strings.Replace(strings.Replace(agentsUrl, "http://", "", 1), "/v1/agent", "", 1), ":")[0],
+			"port": strings.Split(strings.Replace(strings.Replace(agentsUrl, "http://", "", 1), "/v1/agent", "", 1), ":")[1],
```
You could also get the host/port like this:
```go
u, err := url.Parse(ts.URL)
host, port, err := net.SplitHostPort(u.Host)
```

Mierdin replied:
Those are both extremely useful functions to know! Thanks!

vCabbage commented on api/client/agents_test.go:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintln(w, string(agentJson))
```
This could be `w.Write(agentJson)`

Mierdin replied:
Yup. Still using my old way of doing things here

vCabbage commented on api/client/agents_test.go:
```diff
@@ -0,0 +1,95 @@
+	for _, test := range agentTests {
+		err := capi.DisplayAgents(test.arg1, test.arg2)
+		if err != nil {
+			t.Fatal(err)
```
Should this be `t.Error` so that all the tests get run instead of stopping at the first failure?

Mierdin replied:
I personally don't care if it doesn't run all the tests, but others might, so I'll go back to using Error.

vCabbage commented:
Speaking of testing, I came across a slide deck for [Advanced Testing with Go](https://speakerdeck.com/mitchellh/advanced-testing-with-go) by Mitchell Hashimoto. I think there are some very good ideas in there. Supposedly they're going to post a video of the presentation eventually http://www.meetup.com/NewStore/events/227670331/)

vCabbage commented on api/client/agents.go:
```diff
@@ -21,10 +21,12 @@ import (
 // a list of them to the user. Optionally, the user can provide a subargument containing the UUID of
 // a registered agent, and this function will output more detailed information about that agent.
-func (capi ClientApi) Agents(conf map[string]string, agentUuid string) {
+func (capi ClientApi) Agents(conf map[string]string, agentUuid string) (error, []defs.AgentAdvert) {
```
Convention is for the error to be the last return value (ie, `([]defs.AgentAdvert, error)`

Mierdin replied:
Dur! Fixed.

Mierdin commented:
Alright I pushed a commit that addressed your comments. Thanks for those!

And thanks for the reference to the slide deck - that looks like a great talk, and hope we get to see the video! As you can see I have tried to leverage table-driven testing where possible (this was a simple example, hoping to expand on this) and I definitely have a lot of work to do in ToDD wrt writing testable code. That's my singular focus at the moment
